components/launcher-client/src/client.rs

```diff
@@ -75,6 +95,46 @@ impl LauncherCli {
         }
     }
 
+    /// EXPERIMENTAL! BEWARE! EXPERIMENTAL!
+    ///
+    /// When this was added, we needed a way for a Supervisor to be
+    /// able to recover from sending a message to an older version of
+    /// the Launcher that didn't know how to process that message. At
+    /// the time, the Launcher would just ignore the message, but we
+    /// always use a `recv` to wait for the response, so the
+    /// Supervisor would hang.
+    ///
+    /// This function is an attempt to get around this situation, but
+    /// ONLY for the specific scenarios where it is appropriate. I am
+    /// hesitant to introduce a global timeout at this point, because
+    /// it's difficult to know how that will affect the overall
+    /// interactions between the Supervisor and the Launcher (it
+    /// *should* be fine, but I can't guarantee that right now).
+    ///
+    /// As such, use this with caution and intention.
+    fn recv_timeout<T>(rx: &IpcReceiver<Vec<u8>>, timeout: Duration) -> Result<T>
+        where T: protocol::LauncherMessage
+    {
+        // If ipc_channel implemented this directly, we wouldn't have
+        // to do this :(
+        let timeout = Instant::now() + timeout;
+        loop {
+            match rx.try_recv().map_err(|e| Error::from(*e)) {
+                Ok(bytes) => return Self::read(&bytes),
+                Err(Error::IPCIO(io::ErrorKind::WouldBlock)) => {
```

Comment:
This is cool! It is exactly what Rust's async IO implementations do but instead of simply sleeping, it registers a `Waker` that uses underlying syscalls to determine when more data is available.

Reply:
Neat :smile: That sure would be nicer than sleeping.